File: backend/validator/RegisterValidator.php
<?php
// backend/validators/RegisterValidator.php

require_once 'ValidatorBase.php';

class RegisterValidator extends ValidatorBase {
  
  public static function validate($postData): array {
      $validator = new self($postData); // Composición con base
      $errors = [];
      $cleanData = [];

      // ✅ 1. Validar campos individuales de familia
      $familiaFields = [
          'email' => trim($postData['email'] ?? ''),
          'telefono' => trim($postData['telefono'] ?? ''),
          'password' => $postData['password'] ?? '',
          'confirmPassword' => $postData['confirmPassword'] ?? ''
      ];

      $result = $validator->validateFields([
          'email' => $familiaFields['email'],
          'telefono' => $familiaFields['telefono']
      ]);

      $errors = array_merge($errors, $result['errors']);
      $cleanData = array_merge($cleanData, $result['cleanData']);

      // ✅ 2. Validar contraseña con reglas estrictas
      if (!$validator->isValidContrasena($familiaFields['password'])) {
          $errors['password'] = $validator->getErrorMessage('contrasena');
      } else {
          $cleanData['password'] = $familiaFields['password'];
      }

      // ✅ 3. Verificar que las contraseñas coincidan
      if (!$validator->passwordsMatch($familiaFields['password'], $familiaFields['confirmPassword'])) {
          $errors['confirmPassword'] = 'Las contraseñas no coinciden';
      }

      // ✅ 4. Validar datos del administrador
      $adminFields = [
          'adminNombre' => trim($postData['adminNombre'] ?? ''),
          'adminNacimiento' => trim($postData['adminNacimiento'] ?? ''),
          'contraPersonal' => $postData['contraPersonal'] ?? ''
      ];

      if (!$validator->isValidNombre($adminFields['adminNombre'])) {
          $errors['adminNombre'] = $validator->getErrorMessage('nombre');
      } else {
          $cleanData['adminNombre'] = $validator->sanitize($adminFields['adminNombre']);
      }

      if (!$validator->isValidFechaNacimiento($adminFields['adminNacimiento'])) {
          $errors['adminNacimiento'] = $validator->getErrorMessage('fecha_nac');
      } else {
          $cleanData['adminNacimiento'] = $adminFields['adminNacimiento'];
      }

      if (!$validator->isValidContraPersonal($adminFields['contraPersonal'])) {
          $errors['contraPersonal'] = $validator->getErrorMessage('contraPersonal');
      } else {
          $cleanData['contraPersonal'] = $adminFields['contraPersonal'];
      }

      // ✅ 5. Validar miembros opcionales
      if (!empty($postData['miembroNombre']) && is_array($postData['miembroNombre'])) {
        $cleanData['miembros'] = [];
        
        foreach ($postData['miembroNombre'] as $index => $nombre) {
          if (!empty(trim($nombre))) {
            $miembro = [
              'nombre' => trim($nombre),
              'nacimiento' => trim($postData['miembroNacimiento'][$index] ?? ''),
              'contra' => $postData['miembroContra'][$index] ?? '',
              'rol' => trim($postData['miembroRol'][$index] ?? 'miembro')
            ];

            $miembroErrors = [];

            // Validar nombre del miembro
            if (!$validator->isValidNombre($miembro['nombre'])) {
              $miembroErrors[] = "El nombre del miembro #{$index} es inválido";
            }

            // Validar fecha de nacimiento del miembro
            if (!$validator->isValidFechaNacimiento($miembro['nacimiento'])) {
              $miembroErrors[] = "La fecha de nacimiento del miembro '{$miembro['nombre']}' es inválida";
            }

            // Validar contraseña personal del miembro
            if (!$validator->isValidContraPersonal($miembro['contra'])) {
              $miembroErrors[] = "La contraseña del miembro '{$miembro['nombre']}' debe tener 4-6 caracteres";
            }

            // Validar rol del miembro
            if (!$validator->isValidRol($miembro['rol'])) {
              $miembroErrors[] = "El rol del miembro '{$miembro['nombre']}' es inválido";
            }

            if (!empty($miembroErrors)) {
              $errors["miembro_{$index}"] = implode(', ', $miembroErrors);
            } else {
              // Sanitizar y agregar miembro válido
              $cleanData['miembros'][] = [
                'nombre' => $validator->sanitize($miembro['nombre']),
                'nacimiento' => $miembro['nacimiento'],
                'contra' => $miembro['contra'], // No sanitizar contraseñas
                'rol' => strtolower($miembro['rol'])
              ];
            }
          }
        }
      }

      if (!empty($errors)) {
          return [
              'success' => false,
              'errors' => $errors
          ];
      }

      return [
          'success' => true,
          'cleanData' => $cleanData
      ];
  }
}
?>

File: backend/validator/RegisterValidator.go
package validator

import (
	"net/http"
	"regexp"
)

type RegisterValidator struct{}

var RegisterValidatorInstance = &RegisterValidator{}

func (v *RegisterValidator) Validate(r *http.Request) ValidationResult {
	errors := make(map[string]string)

	nombre := r.FormValue("nombre")
	email := r.FormValue("email")
	password := r.FormValue("password")
	confirmPassword := r.FormValue("confirm_password")
	familyName := r.FormValue("family_name")

	// Validar nombre
	if nombre == "" {
		errors["nombre"] = "El nombre es requerido"
	} else if len(nombre) < 2 {
		errors["nombre"] = "El nombre debe tener al menos 2 caracteres"
	}

	// Validar email
	if email == "" {
		errors["email"] = "El email es requerido"
	} else if !isValidEmail(email) {
		errors["email"] = "El formato del email no es válido"
	}

	// Validar contraseña
	if password == "" {
		errors["password"] = "La contraseña es requerida"
	} else if len(password) < 6 {
		errors["password"] = "La contraseña debe tener al menos 6 caracteres"
	}

	// Validar confirmación
	if password != confirmPassword {
		errors["confirm_password"] = "Las contraseñas no coinciden"
	}

	// Validar nombre de familia
	if familyName == "" {
		errors["family_name"] = "El nombre de la familia es requerido"
	}

	return ValidationResult{
		Success: len(errors) == 0,
		Errors:  errors,
	}
}

func isValidEmail(email string) bool {
	emailRegex := `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
	matched, _ := regexp.MatchString(emailRegex, email)
	return matched
}
